pkg/wal/wal.go

```diff
@@ -42,19 +75,759 @@ type Segment interface {
+					bufferVolume = 0
+				}
+				continue
+			case <-timer.C:
+				if bufferVolume == 0 {
+					continue
+				}
+				log.triggerFlushing()
+				bufferVolume = 0
+				continue
+			}
+		}
+	}()
+
+	go func() {
+		log.logger.Info().Msg("Start flush task...")
+
+		defer log.flushWaitGroup.Done()
+
+		for batch := range log.flushChannel {
+			startTime := time.Now()
+
+			var err error
+			for i := 0; i < maxRetries; i++ {
+				if err = log.flushBuffer(batch); err != nil {
+					log.logger.Err(err).Msg("Flushing buffer failed. Retrying...")
+					time.Sleep(100 * time.Millisecond)
+					continue
+				}
+				break
+			}
+			if log.logger.Debug().Enabled() {
+				log.logger.Debug().Msg("Flushed buffer to WAL file. elements: " +
+					strconv.Itoa(batch.count) + ", cost: " + time.Since(startTime).String())
+			}
+
+			batch.notifyRequests(err)
+		}
+		log.logger.Info().Msg("Stop flush task when flush-channel closed!")
+	}()
+
+	log.logger.Info().Msg("Started WAL")
+}
+
+func (log *log) triggerFlushing() {
+	for {
+		select {
+		case log.flushChannel <- log.buffer:
+			if log.logger.Debug().Enabled() {
+				log.logger.Debug().Msg("Send buffer to flush-channel. elements: " + strconv.Itoa(log.buffer.count))
+			}
+			log.newBuffer()
+			return
+		default:
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+func (log *log) newBuffer() {
+	log.buffer.timestampMap = make(map[common.SeriesIDV2][]time.Time)
+	log.buffer.valueMap = make(map[common.SeriesIDV2][]byte)
+	log.buffer.callbackMap = make(map[common.SeriesIDV2][]func(common.SeriesIDV2, time.Time, []byte, error))
+	log.buffer.count = 0
+}
+
+func (log *log) flushBuffer(buffer buffer) error {
+	if buffer.count == 0 {
+		return nil
+	}
+
+	defer func() {
+		log.bytesBuffer.Reset()
+		log.timestampsBuffer.Reset()
+	}()
+
+	// placeholder, preset batch length value is 0
+	batchLen := 0
+	if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int64(batchLen)); err != nil {
+		return errors.Wrap(err, "Write batch length error")
+	}
+	for seriesID, timestamps := range buffer.timestampMap {
+		// Generate seriesID binary
+		seriesIDBytes := seriesID.Marshal()
+		seriesIDBytesLen := len(seriesIDBytes)
+
+		// Generate timestamps compression binary
+		log.timestampsBuffer.Reset()
+		timestampWriter := encoding.NewWriter()
+		timestampEncoder := encoding.NewXOREncoder(timestampWriter)
+		timestampWriter.Reset(log.timestampsBuffer)
+		for _, timestamp := range timestamps {
+			timestampEncoder.Write(timeTouUnixNano(timestamp))
+		}
+		timestampWriter.Flush()
+		timestampsBytes := log.timestampsBuffer.Bytes()
+		timestampsBytesLen := len(timestampsBytes)
+
+		// Generate values compression binary
+		valuesBytes := snappy.Encode(nil, buffer.valueMap[seriesID])
+
+		// Write entry data
+		entryLen := seriesIDLength + seriesIDBytesLen + seriesCountLength + timestampsBinaryLength + timestampsBytesLen + len(valuesBytes)
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int64(entryLen)); err != nil {
```

Comment:
You should provide several binary packing/unpacking methods to conceal bytesBuffer and the binary order.